### PR TITLE
Add test to visit sites with SocketsHttpHandler

### LINKED ISSUE #27792: Consider some search/spider-based testing for SocketsHttpHandler
We could, for example, write a simple client that did some Bing searches and then made requests to the top 1000 results with multiple handlers and validate various things, at a minimum that requests succeed and fail similarly, maybe something about the content if possible, etc. We could also, for example, write a little spider that walked over some sites and did similar validation.

### PR DESCRIPTION
**Pull request by pjanotti**
Adds a small test to visit a list of sites using SocketsHttpHandler and allow easy debugging and comparison with platform handlers in case one needs to investigate. The tests are disabled by default and are controlled by Xunit traits.

Closes #27792

### CODE REVIEW

File: src/System.Net.Http/tests/FunctionalTests/HttpClient.SelectedSitesTest.cs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        }
+
+        [Theory]
+        [OuterLoop]
+        [Trait("SiteInvestigation", "true")]
+        [InlineData("http://microsoft.com")]
+        public async Task RetrieveSite_Debug_Helper(string site)
+        {
+            await VisitSite(site);
+        }
```
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/HttpClient.SelectedSitesTest.cs]: Is this just to validate that VisitSite is working?

File: src/System.Net.Http/tests/FunctionalTests/HttpClient.SelectedSitesTest.cs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+        public static IEnumerable<string[]> GetSelectedSites()
+        {
+            const string resourceName = "SelectedSitesTest.txt";
```
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/HttpClient.SelectedSitesTest.cs]: It's a relatively short list... how about just putting it into an array in the code?

File: src/System.Net.Http/tests/FunctionalTests/HttpClient.SelectedSitesTest.cs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                while (null != (site = reader.ReadLine()))
+                {
+                    yield return new[] { site };
+                }
+            }
+        }
+
+        private async Task VisitSite(string site)
+        {
+            using (HttpClient httpClient = CreateHttpClientForSiteVisit())
```
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/HttpClient.SelectedSitesTest.cs]: Do we need to create a new HttpClient per site?

File: src/System.Net.Http/tests/FunctionalTests/HttpClient.SelectedSitesTest.cs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        private async Task VisitSiteWithClient(string site, HttpClient httpClient)
+        {
+            using (HttpResponseMessage response = await httpClient.GetAsync(site))
+            {
+                switch (response.StatusCode)
+                {
+                    case HttpStatusCode.Redirect:
+                    case HttpStatusCode.OK:
+                        if (response.Content.Headers.ContentLength > 0)
+                            Assert.True((await response.Content.ReadAsByteArrayAsync()).Length > 0);
```
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/HttpClient.SelectedSitesTest.cs]: Should this be an Equal for ContentLength and Length?

File: src/System.Net.Http/tests/FunctionalTests/PlatformHandlerTest.cs [hunk 5]
```diff
@@ -42,6 +42,11 @@ public sealed class PlatformHandler_DiagnosticsTest : DiagnosticsTest
         protected override bool UseSocketsHttpHandler => false;
     }
 
+    public sealed class PlatformHandler_HttpClient_SelectedSites_Test : HttpClient_SelectedSites_Test
+    {
+        protected override bool UseSocketsHttpHandler => false;
+    }
```
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/PlatformHandlerTest.cs]: We're bailing out of the test if UseSocketsHttpHandler is false... do we need this at all then?  Could the test just be moved to a SocketsHttpHandler-specific test class?

File: src/System.Net.Http/tests/FunctionalTests/HttpClient.SelectedSitesTest.cs (hunk 1 above)
Comment (pjanotti) [reply src/System.Net.Http/tests/FunctionalTests/HttpClient.SelectedSitesTest.cs]: This is the one that by default goes for SockesHttpHandler and platform handlers, so if you have one (or a small number) site to compare or debug you just need to update the inline data.

File: src/System.Net.Http/tests/FunctionalTests/HttpClient.SelectedSitesTest.cs (hunk 2 above)
Comment (pjanotti) [reply src/System.Net.Http/tests/FunctionalTests/HttpClient.SelectedSitesTest.cs]: With this I can easily swap to much larger list without changing the sources - that's what I've been doing in my runs (~1500 sites).

Comment (stephentoub) [reply src/System.Net.Http/tests/FunctionalTests/HttpClient.SelectedSitesTest.cs]: Ok

File: src/System.Net.Http/tests/FunctionalTests/HttpClient.SelectedSitesTest.cs (hunk 3 above)
Comment (pjanotti) [reply src/System.Net.Http/tests/FunctionalTests/HttpClient.SelectedSitesTest.cs]: No, I had some runs with a single HttpClient and went back and forth between it and creating a new one. I didn't notice any relevant difference, if you think it is worth I can change to a single HttpClient.

File: src/System.Net.Http/tests/FunctionalTests/HttpClient.SelectedSitesTest.cs (hunk 4 above)
Comment (pjanotti) [reply src/System.Net.Http/tests/FunctionalTests/HttpClient.SelectedSitesTest.cs]: Yes, that will be better.

File: src/System.Net.Http/tests/FunctionalTests/HttpClient.SelectedSitesTest.cs (hunk 3 above)
Comment (stephentoub) [reply src/System.Net.Http/tests/FunctionalTests/HttpClient.SelectedSitesTest.cs]: Either way is fine.  We should be able to reuse it, so I was curious why a new one was being used each time.  What you have is fine, though.

File: src/System.Net.Http/tests/FunctionalTests/PlatformHandlerTest.cs (hunk 5 above)
Comment (pjanotti) [reply src/System.Net.Http/tests/FunctionalTests/PlatformHandlerTest.cs]: That is for the "investigating" test case: if it is interesting somehow I typically wanted to see the behavior of the platform handler (and perhaps debug it too).

**Pull request merged**